week7.2-ex1.py

```diff
@@ -0,0 +1,153 @@
+
+
+class PostOffice:
+    """A Post Office class. Allows users to message each other.
+
+    Parameters
+    ----------
+    usernames : list
+        Users for which we should create PO Boxes.
+
+    Attributes
+    ----------
+    message_id : int
+        Incremental id of the last message sent.
+    boxes : dict
+        Users' inboxes.
+    """
+
+    def __init__(self, usernames):
+        self.message_id = 0
+        self.boxes = {user: [] for user in usernames}
+
+    def send_message(self, sender, recipient, message_body, urgent=False):
+        """Send a message to a recipient.
+
+        Parameters
+        ----------
+        sender : str
+            The message sender's username.
+        recipient : str
+            The message recipient's username.
+        message_body : str
+            The body of the message.
+        urgent : bool, optional
+            The urgency of the message.
+            Urgent messages appear first.
+
+        Returns
+        -------
+        int
+            The message ID, auto incremented number.
+
+        Raises
+        ------
+        KeyError
+            If the recipient does not exist.
+
+        Examples
+        --------
+        After creating a PO box and sending a letter,
+        the recipient should have 1 messege in the
+        inbox.
+
+        >>> po_box = PostOffice(['a', 'b'])
+        >>> message_id = po_box.send_message('a', 'b', 'Hello!')
+        >>> len(po_box.boxes['b'])
+        1
+        >>> message_id
+        1
+        """
+        user_box = self.boxes[recipient]
+        self.message_id = self.message_id + 1
+        message_details = {
+            'id': self.message_id,
+            'body': message_body,
+            'sender': sender,
+            'unread': True
+        }
+        if urgent:
+            user_box.insert(0, message_details)
+        else:
+            user_box.append(message_details)
+        return self.message_id
+
+    def read_inbox(self, username: str, messages_number: int = -1) -> List:
+        """Read the unread messages in the inbox.
+
+        Parameters
+        ----------
+        username : str
+            The recipient's username.
+        messages_number : int, optional
+            The number of the messages to read.
+            If no number are received, returns all unread messages in the inbox.
+
+        Returns
+        -------
+        List
+            The list of the unread messages of the recipient's inbox.
+
+        Raises
+        ------
+        KeyError
+            If the username does not exist.
+        """
+        if messages_number == -1:
+            messages_number = len(self.boxes[username])
+        massages = [self.boxes[username][number] for number in range(messages_number) if
+                    self.boxes[username][number]['unread']]
+        for number in range(messages_number):
+            if self.boxes[username][number]['unread']:
+                self.boxes[username][number]['unread'] = False
```

Comment:
How can you impement this more elegantly with one loop

Reply:
There is only one loop here. Did you mean one line? I do not know how to change in one line, just extract information, I searched the internet and did not find what I wanted.

Reply:
There is this loop and the previous list comprehension obove it, both "loops" can be connected as they run through the same parameters.